test/integration/cls.test.js

```diff
@@ -141,6 +141,199 @@ if (current.dialect.supports.transactions) {
+
+      testHooks({
+        method: 'Model.upsert',
+        hooks: ['beforeUpsert', 'afterUpsert'],
+        optionPos: 1,
+        async execute(User) {
+          await User.upsert({
+            name: 'bob',
+          });
+        },
+      });
+
+      testHooks({
+        method: 'Model.destroy',
+        hooks: ['beforeBulkDestroy', 'afterBulkDestroy'],
+        optionPos: 0,
+        async execute(User) {
+          await User.destroy({ where: { name: 'bob' } });
+        },
+      });
+
+      testHooks({
+        method: 'Model.destroy with individualHooks',
+        hooks: ['beforeDestroy', 'beforeDestroy'],
+        optionPos: 1,
+        async execute(User) {
+          await User.create({ name: 'bob' });
+          await User.destroy({ where: { name: 'bob' }, individualHooks: true });
+        },
+      });
+
+      testHooks({
+        method: 'Model#destroy',
+        hooks: ['beforeDestroy', 'beforeDestroy'],
+        optionPos: 1,
+        async execute(User) {
+          const user = await User.create({ name: 'bob' });
+          await user.destroy();
+        },
+      });
+
+      testHooks({
+        method: 'Model.update',
+        hooks: ['beforeBulkUpdate', 'afterBulkUpdate'],
+        optionPos: 0,
+        async execute(User) {
+          await User.update({ name: 'alice' }, { where: { name: 'bob' } });
+        },
+      });
+
+      testHooks({
+        method: 'Model.update with individualHooks',
+        hooks: ['beforeUpdate', 'afterUpdate'],
+        optionPos: 1,
+        async execute(User) {
+          await User.create({ name: 'bob' });
+          await User.update({ name: 'alice' }, { where: { name: 'bob' }, individualHooks: true });
+        },
+      });
+
+      testHooks({
+        method: 'Model#save (isNewRecord)',
+        hooks: ['beforeCreate', 'afterCreate'],
+        optionPos: 1,
+        async execute(User) {
+          const user = User.build({ name: 'bob' });
+          user.name = 'alice';
+          await user.save();
+        },
+      });
+
+      testHooks({
+        method: 'Model#save (!isNewRecord)',
+        hooks: ['beforeUpdate', 'afterUpdate'],
+        optionPos: 1,
+        async execute(User) {
+          const user = await User.create({ name: 'bob' });
+          user.name = 'alice';
+          await user.save();
+        },
+      });
+
+      // these paranoid test getting failed the transaction is going in the runhook function but still the test is failing
+      describe('paranoid restore', () => {
+        beforeEach(async function () {
+          this.ParanoidUser = this.sequelize.define('paranoid_user', {
+            name: Sequelize.STRING,
+          }, { paranoid: true });
+
+          await this.ParanoidUser.sync({ force: true });
+        });
+
+        testHooks({
+          method: 'Model.restore',
+          hooks: ['beforeBulkRestore', 'afterBulkRestore'],
+          optionPos: 0,
+          async execute() {
+            const User = this.ParanoidUser;
```

Comment:
this test is using `this.ParanoidUser` (and it should), but `testHooks` uses `this.User`

Pass the model to `testHooks` and make `testHooks` use that model instead of `this.User`

```
testHooks({
  getModel: () => this.ParanoidUser,
  // ...
});
```

Note that it has to be a getter like above because `this.ParanoidUser` doesn't exist until `beforeEach` has been executed